subsys/trusted_storage/src/auth_crypt_backend.c

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <zephyr/device.h>
+#include <init.h>
+#include <sys/util.h>
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include <errno.h>
+
+#include "settings_helpers.h"
+#include "auth_crypt_backend.h"
+
+/* AEAD IV suffix */
+#define TRUSTED_STORAGE_FILENAME_SUFFIX_IV		".iv"
+
+/*
+ * AEAD based Authenticated Encrypted trust implementation
+ *
+ * Actual implementation uses:
+ * - Hexadecimal UID imported as KEY
+ * - UID+Flags+Size as additional parameter
+ * - IV is an incrementing number at each encryption
+ * - TAG is left at the end of output data
+ */
+
+/* TODO Kconfig */
+#define AEAD_TAG_SIZE		16
+#define AEAD_IV_SIZE		12
+
+/* Max storage size for the encrypted or decrypted output */
+#define AEAD_MAX_BUF_SIZE ROUND_UP(TRUSTED_STORAGE_MAX_ASSET_SIZE +	\
+				   AEAD_TAG_SIZE, AEAD_TAG_SIZE)
+
+/* 128bits AEAD key pattern: uid low, uid */
+#define AEAD_KEY_PATTERN	"%08x%08x"
+#define AEAD_KEY_SIZE		(sizeof(uint64_t) * 2)
+
+/* Additional data structure */
+struct aead_additional_data {
+	psa_storage_uid_t uid;
+	psa_storage_create_flags_t flags;
+	size_t size;
+};
+
+static uint64_t g_iv_low;
+static uint64_t g_iv_high;
+
+/* Return an incrementing IV */
+static psa_status_t psa_ps_get_iv(uint8_t *iv, size_t iv_len)
+{
+	if (iv_len == 0)
+		return PSA_SUCCESS;
+
+	/* Incrementing is implemented by using 2 64bit numbers */
+	g_iv_low++;
+
+	/* On overflow */
+	if (g_iv_low == 0)
+		++g_iv_high;
+
+	/* Copy low first */
+	memcpy(iv, &g_iv_low, MIN(iv_len, sizeof(g_iv_low)));
+
+	/* Then high is there's free space */
+	if (iv_len > sizeof(g_iv_low)) {
+		memcpy(iv + sizeof(g_iv_low), &g_iv_high,
+				MIN(iv_len - sizeof(g_iv_low), sizeof(g_iv_high)));
+	}
+
+	return PSA_SUCCESS;
+}
+
+/* Temporary AEAD encryption/decryption buffers */
+static uint8_t aead_buf[AEAD_MAX_BUF_SIZE];
+static uint8_t data_buf[TRUSTED_STORAGE_MAX_ASSET_SIZE];
+
+psa_status_t trusted_storage_auth_crypt_backend_init(
+					psa_status_t (*init_cb)(void),
+					psa_status_t (*get_random_cb)(
+						void *buf, size_t length))
+{
+	psa_status_t status;
+
+	if (init_cb) {
+		status = init_cb();
+		if (status != PSA_SUCCESS) {
+			return -EIO;
+		}
+	}
+
+	if (get_random_cb) {
```

Comment:
The unpacking of the random data is noisy here. If the marked API is get_iv when the data is needed, it will increase the cohesion and limit tight coupling.